### ISSUE TITLE
C-based pypi package

### ISSUE THREAD

**Issue by ThomasWaldmann** [oconnor663/blake3-py#30]
It would be cool to have a C-based pypi package for the blake3 bindings.

I've read the discussion on the Python issue tracker about hashlib inclusion, but I guess that might take a while or might never happen.

OTOH, I guess there are some projects out there (like `borgbackup`) that would like to add blake3, but not introduce a dependency on the rust toolchain. The latter might be no big issue if you target modern linux dists, but if you try to support a lot of platforms, this at least creates a lot of uncertainty (that block adoption) if not real issues (if one adopts and then stumbles into them).

Also, considering python 3.10 is quite new still, it would be cool if the package also worked for python 3.9.

**Comment by ThomasWaldmann** [comment]
Guess this is needed to be compatible with py39+:
```
diff --git a/c_impl/blake3module.c b/c_impl/blake3module.c
index f4160fb..0fce6ad 100644
--- a/c_impl/blake3module.c
+++ b/c_impl/blake3module.c
@@ -414,7 +414,8 @@ PyMODINIT_FUNC PyInit_blake3(void) {
     goto exit;
   }
 
-  if (PyModule_AddObjectRef(module, "blake3", (PyObject *)&Blake3Type) < 0) {
+  if (PyModule_AddObject(module, "blake3", (PyObject *)&Blake3Type) < 0) {
+    Py_DECREF((PyObject *)&Blake3Type);
     goto exit;
   } 
```

**Comment by ddelange** [comment]
Hi @ThomasWaldmann :wave:

fyi, blake3 v1.0 has been released with a stable API, and with [lots of wheels](https://pypi.org/project/blake3/1.0.1/#files) to cover a modern audience.

If you're missing a specific wheel there, please give a shout. I don't think the reference C implementation was ever intended for production.

Hope that helps!

**Comment by ThomasWaldmann** [comment]
Thanks for the pointer. Yeah, that's quite a lot already.

Missing OSes:
- *BSD
- Solaris derivates, like e.g. OpenIndiana

Not sure about the situation on windows. Would such a windows wheel also work e.g. on cygwin?

**Comment by ThomasWaldmann** [comment]
Another data point:

Some years passed since I opened this ticket and availability of and dependency on rust have changed.

E.g. I recently had to add rust toolchain to my *BSD platform testing machines, because a borg master branch dependency (not blake3) required it.

**Comment by ddelange** [comment]
> PEP 600 has been designed to be "future-proof" and does not enforce specific symbols and a specific distro to build. It only states that a wheel tagged `manylinux_x_y` shall work on any distro based on `glibc>=x.y`. PEP 656 added `musllinux_x_y` tags for `musl>=x.y`

ref https://github.com/pypa/manylinux 

concretely, the blake3 wheels will install on any linux distro (incl BSD etc) which have glibc>=2.17 or musl>=1.1.

regarding cygwin I think [this comment](https://github.com/pypa/cibuildwheel/issues/329#issuecomment-1152740563) explains why you don't see cygwin wheels in the wild. my bet would be that modern users will either be installing the windows wheels natively (using python compiled for windows), or installing the linux wheels using the WSL (https://superuser.com/questions/1360748).

**Comment by oconnor663** [comment]
> I don't think the reference C implementation was ever intended for production.

Yeah the main purpose of the C implementation was as a proof-of-concept for including BLAKE3 in the Python standard library (which already has BLAKE2). If I remember correctly, they ultimately decided they didn't want to take it until there was OpenSSL support.

**Comment by datumbox** [comment]
Hello @oconnor663 , thanks a lot for the great work on the project.

I would also love to get my hands on some C-based pypi package. I'm working on a side project (nothing serious/professional) where I'm investigating the use of Blake3 for keystream generation. I've a PR (see https://github.com/datumbox/VernamVeil/pull/16) where I vendor the reference C implementation but it's a pain to setup for all the hardware accelerations, especially if I don't have access to the hardware.

I saw you have a setup.py file in the `c_impl` folder that could do the trick but in general I hate grabbing source from other projects and I would rather link to your C package if available (I really need C, can't do it via python). If that's not within your plans to provide any time soon, could you clarify what would be your advised way for reusing that part in my project? Would you even advise it (I saw some notes that the implementation is far from perfect but who can claim perfection anyway)? Thanks in advance!

**Comment by oconnor663** [comment]
@datumbox could you tell me more about why you don't want to depend on the published Rust implementation?

**Comment by datumbox** [comment]
hey @ddelange!

> you want blake3-py as a pip dependency

Not exactly. The blake3-py can indeed be one of the dependencies as a drop-in replacement of hashlib.blake2b for Seed evolution, synthetic IV creation and general hashing in the cypher. Nevertheless I also have a mechanism that is used on the keystream generation of the cypher that uses Counts stored in a Numpy array + the current seed, to generate bytes. This is written in C for blake2b and this is where I was considering integrating with Blake3's official C implementation. I have that already at https://github.com/datumbox/VernamVeil/pull/16 but with no acceleration. If there was an official C binary, (similar to how OpenSSL provides for blake2b) I would be just using that in my thin wrapper. 

So in this thread, I came to inquiry about the second use-case. The first use-case (general hashing) is indeed covered 100% by the blake3-py.

**Comment by datumbox** [comment]
I've attempted to write  a [parallel tree hashing implementation in C](https://github.com/datumbox/VernamVeil/blob/1844183ac080a8d2fa31ba184971abef5dbcd04b/nphash/_npblake3.c#L72), "inspired" by how it works in the official Rust implementation. Unfortunately faithful reproduction is not possible because the public API of the C lib doesn't expose low-level APIs for direct manipulation of chunk counters, flags, etc. As a result we don't support domain separation for chunks, parent node compression and faithful root finalisation.

I **think** the approach works for my use-case but I wouldn't use the implementation anywhere near production. The only reason I'm updating the post is incase on the maintainers want to take a look and perhaps eventually offer an official implementation for C users.

**Comment by oconnor663** [comment]
I'm confused by these parts:

> I would rather link to your C package if available (I really need C, can't do it via python)

> This is written in C for blake2b and this is where I was considering integrating with Blake3's official C implementation.

It sounds like you're writing C code. How does that relate to a PyPI package? Is it clear why I'm confused? Please help me catch up.

Separately, we did recently add multithreading support to the BLAKE3 C implementation. Have you seen this? https://github.com/BLAKE3-team/BLAKE3/tree/master/c#multithreading

**Comment by datumbox** [comment]
Thanks for clarifying and sorry for the confusion!

Yes, I’m writing C code (with a Python wrapper), and my ideal scenario would be to have an official C-based PyPI package that ships pre-built shared libraries (.so/.dylib/.dll) with all hardware accelerations enabled (NEON, AVX, etc). My hope was that I could then link against that shared library from my own C extension, rather than vendoring or building the C code myself.

I wasn't aware about the recent addition of multithreading support on C. I guess, that's what I'll be doing next. Thanks a lot for the pointer.

**Comment by ThomasWaldmann** [comment]
A small update from borgbackup (master branch, borg2 future release):

Looks like we can get `blake3` pypi package (and rust, when there is no binary wheel) working on most platforms. The only platform I couldn't get it working is Haiku OS (because their rust is broken).

But Linux, macOS, *BSD, omniOS ("Open Solaris"), Windows/msys2 works. \o/

https://github.com/borgbackup/borg/pull/9643

Also benchmark on Apple M3 Pro looked really good:
```
hmac-sha256              1.00 GB    0.327s  <-- with hw accelerated sha256
blake2b-256              1.00 GB    0.952s  <-- pure sw, used in borg 1.x
blake3                   1.00 GB    0.442s  <-- pure sw, single-threaded
blake3-mt                1.00 GB    0.078s  <-- pure sw, with MT.
```

Currently, I use `blake3>=1.0.0` as requirement, is that OK or was there some notable issue in specific 1.0.x releases I should better exclude?

The plan is to keep legacy blake2b support only for importing borg 1.x repos, but only offer blake3 for borg2 repos.
